pkg/virt-operator/kubevirt.go

```diff
@@ -465,53 +462,87 @@ func (c *KubeVirtController) syncDeployment(kv *v1.KubeVirt) error {
 
 	if objectsAdded == 0 {
-		// deployment successful
-		kv.Status.Phase = v1.KubeVirtPhaseDeployed
 
-		// remove old conditions, add Created condition
-		kv.Status.Conditions = []v1.KubeVirtCondition{}
+		// add Created condition
 		util.UpdateCondition(kv, v1.KubeVirtConditionCreated, k8sv1.ConditionTrue, ConditionReasonDeploymentCreated, "All resources were created.")
-
 		logger.Info("All KubeVirt resources created")
-		return nil
+
+		// check if components are ready
+		if c.isReady() {
+			logger.Info("All KubeVirt components ready")
+			kv.Status.Phase = v1.KubeVirtPhaseDeployed
+			util.UpdateCondition(kv, v1.KubeVirtConditionReady, k8sv1.ConditionTrue, ConditionReasonDeploymentReady, "All components are ready.")
+			return nil
+		}
+		util.RemoveCondition(kv, v1.KubeVirtConditionReady)
+
+	} else {
+		util.RemoveCondition(kv, v1.KubeVirtConditionCreated)
 	}
 
 	logger.Info("Processed deployment for this round")
 	return nil
 }
 
+func (c *KubeVirtController) isReady() bool {
+
+	for _, obj := range c.stores.DeploymentCache.List() {
+		if deployment, ok := obj.(*appsv1.Deployment); ok {
+			var specReplicas int32 = 1
+			if deployment.Spec.Replicas != nil {
+				specReplicas = *deployment.Spec.Replicas
+			}
+			if specReplicas == deployment.Status.Replicas &&
+				deployment.Status.Replicas != deployment.Status.ReadyReplicas {
+				log.Log.V(4).Infof("Deployment %v not ready yet", deployment.Name)
+				return false
+			}
+		}
+	}
+
+	for _, obj := range c.stores.DaemonSetCache.List() {
+		if daemonset, ok := obj.(*appsv1.DaemonSet); ok {
+			if daemonset.Status.DesiredNumberScheduled == 0 ||
+				daemonset.Status.DesiredNumberScheduled != daemonset.Status.NumberReady {
+
+				log.Log.V(4).Infof("DaemonSet %v not ready yet", daemonset.Name)
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
 func (c *KubeVirtController) syncDeletion(kv *v1.KubeVirt) error {
 	logger := log.Log.Object(kv)
 	logger.Info("Handling deletion")
 
-	// delete
-	if kv.Status.Phase == v1.KubeVirtPhaseDeleted {
-		logger.Info("Is already deleted")
-		return nil
-	}
-
 	// set phase to deleting
 	kv.Status.Phase = v1.KubeVirtPhaseDeleting
 
+	// remove created and ready conditions
+	util.RemoveCondition(kv, v1.KubeVirtConditionCreated)
+	util.RemoveCondition(kv, v1.KubeVirtConditionReady)
+
 	err := deletion.Delete(kv, c.clientset, c.stores, &c.kubeVirtExpectations)
 	if err != nil {
 		// deletion failed
 		util.UpdateCondition(kv, v1.KubeVirtConditionSynchronized, k8sv1.ConditionFalse, ConditionReasonDeletionFailedError, fmt.Sprintf("An error occurred during deletion: %v", err))
 		return err
 	}
+	util.RemoveCondition(kv, v1.KubeVirtConditionSynchronized)
 
 	if c.stores.AllEmpty() {
 
 		// deletion successful
 		kv.Status.Phase = v1.KubeVirtPhaseDeleted
 
-		// remove conditions
-		kv.Status.Conditions = []v1.KubeVirtCondition{}
-
 		// remove finalizer
-		kv.Finalizers = []string{}
+		kv.Finalizers = nil
```

Comment:
out of curiosity, does setting Finalizers to nil vs. []string{} change behavior? For my understanding, i'm wondering why this change was made.

Reply:
in an earlier iteration I had this statement at another place, and it triggered a KubeVirt CR update, when finalizers on a new CR were nil, and I replaced it with an empty array :)